EZP-31126: Add more Online Editor use cases (#942)
ezsystems/developer-documentation: docs/guide/extending/extending_online_editor.md

![FactBox Tag](img/custom_tag_factbox.png "FactBox Tag in the Online Editor" )

### Inline custom tags

Custom tags can also be placed inline with the following configuration:

``` yaml hl_lines="6"
ezrichtext:
    custom_tags:
        badge:
            template: field_type/ezrichtext/custom_tag/badge.html.twig
            icon: '/bundles/ezplatformadminui/img/ez-icons.svg#bookmark'
            is_inline: true
            attributes:
                # ...
```

`is_inline` is an optional key.
The default value is `false`, so if it is not set, the custom tag will be treated as a block tag.

!!! caution "Incorrect configuration"

    Newer configuration options, such as `is_inline`, only work with the configuration provided above.

>     If your project uses [old configuration](../../updating/4_update_2.4.md#changes-to-custom-tags),

Reviewer: ```suggestion
    If your project uses [configuration from version prior to 2.4](../../updating/4_update_2.4.md#changes-to-custom-tags),
```

![Link Tag](img/custom_tag_link.png "Link Tag in the Online Editor") 

#### Acronym

You can create an inline custom tag that will display a hovering tooltip with an explanation of an acronym.

``` yaml
ezpublish:
    system:
        admin_group:
            fieldtypes:
                ezrichtext:
                    custom_tags: [acronym]

ezrichtext:
    custom_tags:
        acronym:
            template: field_type/ezrichtext/custom_tag/acronym.html.twig
            icon: '/bundles/ezplatformadminui/img/ez-icons.svg#information'
            is_inline: true
            attributes:
                explanation:
                    type: 'string'
```

The `explanation` attribute will contain the meaning of the acronym that will be provided
while editing in the Online Editor.

Label translations can be provided in `translations/custom_tags.en.yaml`:

``` yaml
ezrichtext.custom_tags.acronym.label: 'Acronym'
ezrichtext.custom_tags.acronym.attributes.meaning.label: 'Explanation'
```

![Adding an explanation to an Acronym custom tag](img/oe_custom_tag_add_acronym.png)

> In the template file `acronym.html.twig` you provide the explanation as `attr_value`

Reviewer: ```suggestion
In the template file `acronym.html.twig` provide the explanation as `attr_value`
```

#### Acronym

You can create an inline custom tag that will display a hovering tooltip with an explanation of an acronym.

``` yaml
ezpublish:
    system:
        admin_group:
            fieldtypes:
                ezrichtext:
                    custom_tags: [acronym]

ezrichtext:
    custom_tags:
        acronym:
            template: field_type/ezrichtext/custom_tag/acronym.html.twig
            icon: '/bundles/ezplatformadminui/img/ez-icons.svg#information'
            is_inline: true
            attributes:
                explanation:
                    type: 'string'
```

The `explanation` attribute will contain the meaning of the acronym that will be provided
while editing in the Online Editor.

Label translations can be provided in `translations/custom_tags.en.yaml`:

``` yaml
ezrichtext.custom_tags.acronym.label: 'Acronym'
ezrichtext.custom_tags.acronym.attributes.meaning.label: 'Explanation'
```

![Adding an explanation to an Acronym custom tag](img/oe_custom_tag_add_acronym.png)

In the template file `acronym.html.twig` you provide the explanation as `attr_value`
to the title of the `abbr` tag:

``` html+twig
{% for attr_value in params %}

>     <abbr title="{{ attr_value }}">{{ content }}</abbr>

Reviewer 2: ```suggestion
    <abbr title="{{ params.explanation }}">{{ content }}</abbr>
```

```

The indicated `note_box.html.twig` template wraps the content of the selected text (`{{ content }}`)
in a custom CSS class:

``` html+twig
<div class="note">{{ content }}</div>
```

``` css
.note {
    display: block;
    background-color: #faa015;
    border-left: solid 5px #353535;
    line-height: 18px;
    padding: 15px;
    color: #fff;
    font-weight: bold;
}
```

Label translation can be provided in `translations/custom_styles.en.yaml`:

``` yaml
ezrichtext.custom_styles.note_box.label: 'Note box'
```

![Adding a Note box custom style](img/oe_custom_style_note_box_select.png)

!!! tip

    You can also create a similar note box using [custom classes](#note-box_1).

#### Text highlight

You can create an inline custom style that highlights a part of a text:

![Example of a custom style highlighting a portion of text](img/oe_custom_style_highlight.png)

``` yaml
ezpublish:
    system:
        admin_group:
            fieldtypes:
                ezrichtext:
                    custom_styles: [highlight]
                    
ezrichtext:
    custom_styles:
        highlight:
            template: field_type/ezrichtext/custom_style/highlight.html.twig

>             inline: true

Reviewer 2: ```suggestion
            is_inline: true
```

    --dir=./templates --output-dir=./translations/ --output-format=yaml
```

### Use cases

#### Note box

You can create a custom class that will enable you to place a paragraph element in a note box:

![Example of a note box custom style](img/oe_custom_style_note_box.png)

``` yaml
ezpublish:
    system:
        admin_group:
            fieldtypes:
                ezrichtext:
                    classes:
                        paragraph:
                            choices: [regular, tip_box, warning_box]
                            default_value: regular
                            required: false
                            multiple: false
```

> This enables you to choose one of three classes for each paragraph element: `regular`, `tip_box`, or `warning_box`

Reviewer: ```suggestion
This enables you to choose one of the following classes for each paragraph element: `regular`, `tip_box`, or `warning_box`
```
